OpenRA.Mods.Common/Traits/BotModules/Squads/States/GroundStates.cs

```diff
@@ -68,57 +84,115 @@ public void Tick(Squad owner)
+
 		public void Activate(Squad owner) { }
 
 		public void Tick(Squad owner)
 		{
+			// Basic check
 			if (!owner.IsValid)
 				return;
 
 			if (!owner.IsTargetValid)
 			{
-				var closestEnemy = FindClosestEnemy(owner);
-				if (closestEnemy != null)
-					owner.TargetActor = closestEnemy;
+				var randomSuitableEnemy = GetRandomValuableTarget(owner);
+				if (randomSuitableEnemy != null)
+					owner.TargetActor = randomSuitableEnemy;
 				else
 				{
 					owner.FuzzyStateMachine.ChangeState(owner, new GroundUnitsFleeState(), true);
 					return;
 				}
 			}
 
-			var leader = owner.Units.ClosestTo(owner.TargetActor.CenterPosition);
-			if (leader == null)
+			// Initialize pathGuide. Optimaze pathfinding by using pathGuide.
+			var pathGuide = owner.Units.FirstOrDefault();
+			if (pathGuide == null)
 				return;
 
-			var ownUnits = owner.World.FindActorsInCircle(leader.CenterPosition, WDist.FromCells(owner.Units.Count) / 3)
-				.Where(a => a.Owner == owner.Units.First().Owner && owner.Units.Contains(a)).ToHashSet();
+			// 1. Threat scan surroundings
+			var attackScanRadius = WDist.FromCells(owner.SquadManager.Info.AttackScanRadius);
 
-			if (ownUnits.Count < owner.Units.Count)
+			var targetActor = ThreatScan(owner, pathGuide, attackScanRadius);
+			if (targetActor != null)
 			{
-				// Since units have different movement speeds, they get separated while approaching the target.
-				// Let them regroup into tighter formation.
-				owner.Bot.QueueOrder(new Order("Stop", leader, false));
-				foreach (var unit in owner.Units.Where(a => !ownUnits.Contains(a)))
-					owner.Bot.QueueOrder(new Order("AttackMove", unit, Target.FromCell(owner.World, leader.Location), false));
+				owner.TargetActor = targetActor;
+				owner.FuzzyStateMachine.ChangeState(owner, new GroundUnitsAttackState(), true);
+				return;
 			}
-			else
+
+			// 2. Force scattered for pathGuide if needed
+			if (StuckInPath <= 0)
 			{
-				var enemies = owner.World.FindActorsInCircle(leader.CenterPosition, WDist.FromCells(owner.SquadManager.Info.AttackScanRadius))
-					.Where(owner.SquadManager.IsPreferredEnemyUnit);
-				var target = enemies.ClosestTo(leader.CenterPosition);
-				if (target != null)
+				if (TryMakeWay > 0)
 				{
-					owner.TargetActor = target;
-					owner.FuzzyStateMachine.ChangeState(owner, new GroundUnitsAttackState(), true);
+					owner.Bot.QueueOrder(new Order("AttackMove", pathGuide, Target.FromCell(owner.World, owner.TargetActor.Location), false));
+
+					foreach (var a in owner.Units)
+					{
+						if (a != pathGuide)
+							owner.Bot.QueueOrder(new Order("Scatter", a, false));
+					}
+
+					TryMakeWay--;
 				}
 				else
-					foreach (var a in owner.Units)
-						owner.Bot.QueueOrder(new Order("AttackMove", a, Target.FromCell(owner.World, owner.TargetActor.Location), false));
+				{
+					// When going through is over, restore the check
+					StuckInPath = StuckInPathCheckTimes + MakeWayTick;
+					TryMakeWay = MakeWayTick;
+				}
```

Comment:
yeah but in this PR I don't want to consider it too complex, and even if the entire squad is stuck, the PF won't lag the game since only one unit doing a big PF, which is at least much better than the original.

Reply:
You can add your method here maybe after this PR is done. I am tired to add new function after that performance issue. Even now the bot_tick can be a little bit high.

Reply:
And this stuck check is only to solve when `PathGuider` is stuck by squad members, which can happen easier than other stuck condition.